toolbox/errorfinder/main.go

```diff
@@ -0,0 +1,394 @@
+							newLine := strings.Replace(line, ":", "<", 1)
+							if strings.Index(newLine, ":") != -1 {
+								indColon := strings.Index(newLine, ":")
+								line = line[:indColon]
+							}
+						} else {
+							line = line[:indexColon]
+						}
+					}
+					errorLines = append(errorLines, line)
+
+					if strings.Index(line, "warn") != -1 {
+						warningPrs := []string{}
+						if warningMap[line] == nil {
+							warningPrs = []string{filePath}
+						} else {
+							warningPrs = warningMap[line]
+							if !contains(warningPrs, filePath) {
+								warningPrs = append(warningPrs, filePath)
+							}
+						}
+						warningMap[line] = warningPrs
+					} else {
+						prSlice := []string{}
+						if errorMap[line] == nil {
+							prSlice = []string{filePath}
+						} else {
+							prSlice = errorMap[line]
+							if !contains(prSlice, filePath) {
+								prSlice = append(prSlice, filePath)
+							}
+						}
+						errorMap[line] = prSlice
+					}
+				}
+			}
+
+			errorString = errorString + "\n" + strings.Join(errorLines, "\n")
+			processedString = splitSlice[1]
+		}
+		contentMap[filePath] = errorString
+	}
+	return contentMap, errorMap, warningMap, nil
+}
+
+func (f *ErrorFinder) generalizeDigits(content string) string {
+	reg, _ := regexp.Compile("[0-9]+")
+	newContent := reg.ReplaceAllString(content, "\\d")
+	reg, _ = regexp.Compile("\\\"(.*?)\\\"")
+	newContent = reg.ReplaceAllString(newContent, "\\s")
+	reg, _ = regexp.Compile("\\'(.*?)\\'")
+	newContent = reg.ReplaceAllString(newContent, "\\s")
+	return newContent
+}
+
+func writeCSV(errorMap map[string][]string, warningMap map[string][]string, fileName string) {
+	file, err := os.Create(fileName)
+	if err != nil {
+		log.Fatal("cannot open file", err)
+	}
+
+	defer file.Close()
+
+	writer := csv.NewWriter(file)
+	defer writer.Flush()
+	error := []string{"error"}
+	err = writer.Write(error)
+	if err != nil {
+		log.Fatal("cannot write line", err)
+	}
+	for key, value := range errorMap {
+		newLine := []string{}
+		newLine = append(newLine, key)
+		s := strconv.Itoa(len(value))
+		newLine = append(newLine, s)
+		newLine = append(newLine, value...)
+		err := writer.Write(newLine)
+		if err != nil {
+			log.Fatal("cannot write line", err)
+		}
+
+	}
+	warning := []string{"warning"}
+	err = writer.Write(warning)
+	if err != nil {
+		log.Fatal("cannot write line", err)
+	}
+	for key, value := range warningMap {
+		newLine := []string{}
+		newLine = append(newLine, key)
+		s := strconv.Itoa(len(value))
+		newLine = append(newLine, s)
+		newLine = append(newLine, value...)
+		err := writer.Write(newLine)
+		if err != nil {
+			log.Fatal("cannot write line", err)
+		}
+
+	}
+}
+
+func main() {
+	var fileName string
+	flag.StringVar(&fileName, "OutputFileName", "", "A file name for output csv file with .csv at the end.")
+	var spreadsheetID string
+	flag.StringVar(&spreadsheetID, "SpreadsheetId", "", "Spreadsheet ID for the spreadsheet containing file path")
```

Comment:
this seems odd to me... Where does the spreadsheet come from? If we have a spreadsheet that contains all of the failures then clearly we have some code that makes this. Why can't we just do this logic there?

Reply:
The spreadsheet contains a list of prs from istio-prow not all of the failures.

Reply:
Where did the spreadsheet come from though? I assume someone didn't make it by hand?

Reply:
Maria shared with me. I don't think she made it by hand. https://docs.google.com/spreadsheets/d/1XK4Nbq7nwqwXMC7waMsEqhe_wOf1J1rqwbECrop9fQI/edit?ts=5d277e24#gid=0

Reply:
Yes, so what I am saying is: Someone wrote code to go through all of our tests and find failures and create this spreadsheet. Now you are creating code to read the spreadsheet and then go through all of our tests and read all of the failures.. Why don't we just have one piece of code that does all of this?

Also @therealmitchconnors already has a bunch of code that parses and categorizes the errors, I would sync up with him to avoid all of this code duplication

Reply:
I asked @therealmitchconnors if he needs help with categorizing the errors and that's what she is doing here. Finding common errors, that later we can identified what is common failures.